src/modules/roc_ctl/task_queue.cpp

```diff
@@ -157,14 +157,25 @@ void TaskQueue::stop_and_wait() {
 
 bool TaskQueue::process_tasks_() {
     core::Mutex::Lock lock(task_mutex_);
+    bool fetch_ready_ = true;
 
     for (;;) {
-        Task* task = fetch_ready_task_();
+        Task* task = NULL;
 
-        if (!task) {
+        if (fetch_ready_) {
+            task = fetch_ready_task_();
+            if (!task) {
+                task = fetch_sleeping_task_();
+            }
+        } else {
             task = fetch_sleeping_task_();
+            if (!task) {
+                task = fetch_ready_task_();
+            }
         }
 
+        fetch_ready_ = !fetch_ready_;
```

Comment:
There is a minor issue here. We're inverting fetch_ready_ no matter of what type of task we have processed actually, ready or sleeping (both types are possible since if one queue is empty, we check another).

Thus, the following situation is possible:

* FIRST CALL: fetch_ready_ is true, ready queue is empty, sleeping queue is non-empty; we process sleeping task and set fetch_ready_ to false

* new tasks is added to ready queue

* SECOND CALL: fetch_ready_ is false, ready queue is non-empty, sleeping queue is non-empty; we process sleeping task again

So, we will process two sleeping tasks in a row, although ready queue is non-empty.

Not a big deal, but I think the algorithm would be a bit clearer if we took into account what type of task we have just processed and always process different type of task on next invocations, it they are present.

Reply:
BTW it means that it's indeed better to make the flag field instead of var. But it doesn't have to be atomic, since it's accessed only from ctl thread.

Reply:
I added else blocks after the task null checks so that the bool only flips on successful fetches